core/src/main/java/hudson/security/PermissionRegistry.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package hudson.security;
+
+import jenkins.model.Jenkins;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Declarative permission registry used for managing which permissions are known by Jenkins.
+ *
+ * @since TODO
+ */
+@ParametersAreNonnullByDefault
+public enum PermissionRegistry {
```

Comment:
Nonstandard way of making a singleton.

Also you probably need to make this instance be tied to `Jenkins.get()`, so that for example tests using `RestartableJenkinsRule` will properly get a new registry after the restart. In general, you cannot use mutable `static` fields (which this effectively is) anywhere in Jenkins other than `Jenkins` itself.

(yes this may already have been broken in `PermissionGroup`, but if so, now is the time to fix it)

Reply:
Nonstandard way of doing a singleton in Jenkins, perhaps, but an `enum` is the most effective singleton in Java due to guarantees by Java that it will be a true singleton regardless of multithreading, initialization ordering, etc.

I remember seeing an error when trying to use an extension point instead for the permission registry, but I don't recall what it was. I'll try that approach again and get back to you.

Reply:
I tried making this an `@Extension` instead of an enum, but now I get an error in the clinit of Jenkins:

```
Oct 25, 2018 11:47:37 AM hudson.WebAppMain contextInitialized
SEVERE: Failed to initialize Jenkins
java.lang.ExceptionInInitializerError
	at jenkins.model.Jenkins.<clinit>(Jenkins.java:5141)
	at hudson.WebAppMain.installLogger(WebAppMain.java:294)
	at hudson.WebAppMain.contextInitialized(WebAppMain.java:128)
	at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:952)
	at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:558)
	at org.eclipse.jetty.server.handler.ContextHandler.startContext(ContextHandler.java:917)
	at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:370)
	at org.eclipse.jetty.webapp.WebAppContext.startWebapp(WebAppContext.java:1497)
	at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1459)
	at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:847)
	at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:287)
	at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:545)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:138)
	at org.eclipse.jetty.server.Server.start(Server.java:416)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:108)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:113)
	at org.eclipse.jetty.server.Server.doStart(Server.java:383)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68)
	at winstone.Launcher.<init>(Launcher.java:184)
	at winstone.Launcher.main(Launcher.java:354)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at Main._main(Main.java:344)
	at Main.main(Main.java:160)
Caused by: java.lang.IllegalStateException: Expected 1 instance of hudson.security.PermissionRegistry but got 0
	at hudson.ExtensionList.lookupSingleton(ExtensionList.java:450)
	at hudson.security.PermissionRegistry.getInstance(PermissionRegistry.java:50)
	at hudson.security.PermissionGroup.register(PermissionGroup.java:161)
	at hudson.security.PermissionGroup.<init>(PermissionGroup.java:78)
	at hudson.security.PermissionGroup.<init>(PermissionGroup.java:64)
	at hudson.security.Permission.<clinit>(Permission.java:297)
	... 27 more
```

Do I need to self-register? I changed the class to extend `ExtensionPoint` and annotated it with `@Extension`, but I'm not sure if it's loaded yet by the time the Jenkins class is being loaded.

Reply:
Doesn't seem like it's possible to use an `ExtensionList` for an extension required by `Jenkins` due to initialization order. This could potentially use a MetaInf type service or something else lower level that executes before `Jenkins` is initialized.